gemma/gemma.cc

```diff
@@ -745,34 +745,23 @@ HWY_NOINLINE void Attention(size_t batch_start, size_t batch_idx, size_t layer,
 
   if constexpr (kHeads == kKVHeads) {
     // Multi-Head Attention
+    static_assert(TConfig::kInterleaveQKV);
+
+    float* HWY_RESTRICT qkv =
+        activations.q.data() + batch_idx * kHeads * kQKVDim * 3;
+    MatVec<kHeads * kQKVDim * 3, kModelDim>(
+        layer_weights->qkv_einsum_w, 0, x, activations.even_odd.data(), qkv,
+        pool);
+
     pool.Run(0, kHeads, [&](const uint64_t head, size_t thread) HWY_ATTR {
```

Comment:
Now that MatVec is outside the parallel section, I am not sure it is still worthwhile. Want to try a regular loop here?

Reply:
The regular loop is slower because Attn() still has the Q*K dot product calculations that get longer and longer. It starts out at roughly the same speed but after 1600 tokens it ends up half as fast.